### ISSUE TITLE
getCurrentFreq() does not return the actual frequency

### ISSUE THREAD

**Issue by GenghisChen** [oshi/oshi#966]
Windows 10
Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.3+7)
oshi 4.0.0

When reading getCurrentFreq(), the values returned are always:
3192000000, 3192000000, 3192000000, 3192000000, 3192000000, 3192000000, 3192000000, 3192000000, 3192000000, 3192000000, 3192000000, 3192000000

These values are similar to the ones received from getVendorFreq() and getMaxFreq(), and do not change when the CPU load changes.
In those cases, SpeedStep may brings down the frequency to  799MHz and turbo brings it up to 4.3GHz (As reported by CPU-Z, OpenHardwareMonitor etc.), but OSHI reports the same values

Please let me know if there's any additional information i can supply

P.S. I expected to see only 6 values in that array, AFAIK when Hyper-Threading is involved, temperature and frequency are determined per physical core, not logical core

**Comment by dbwiddis** [comment]
Welcome to the world of trying to find the right information and failing. :)

OSHI gets the current frequency values on Windows from a native call to [`CallNtPowerInformation`](CallNtPowerInformation) with the `ProcessorInformation` information level. This returns a [`PROCESSOR_POWER_INFORMATION`](https://docs.microsoft.com/en-us/windows/win32/power/processor-power-information-str) structure "for each processor installed on the system." This appears to count logical processors, thus the decision to return the array as-is (although it fails in the case of Processor Groups).  In practicality it appears to return the same value for every processor so it doesn't matter which one.

Oshi returns the "CurrentMhz" value from that native call which the documents indicate is "the maximum specified processor clock frequency multiplied by the current processor throttle."  As you see, this value is not always reliable.  I've also looked at using WMI to query the [Win32_Processor](https://docs.microsoft.com/en-us/windows/win32/cimwin32prov/win32-processor) class, where the "CurrentClockSpeed" field is supposed to provide the same information (but only returns a single value), but this appears to be the same value.

When I initially implemented this (see #418 and implemented in #779 in [this commit](https://github.com/oshi/oshi/pull/779/commits/442a7c59172762358993b27a814473caea8ddf06)) I searched around for alternatives and found none, other than actually measuring clock cycles over a period of time  using a high frequency counter.  This may be what programs like CPU-Z or OpenHardwareMonitor produce, and it's something I've thought about but not had the time to implement.

I'm open to any ideas of other places to retrieve that information from Windows.

**Comment by dbwiddis** [comment]
It looks like external programs use the CPUID opcode to [read IA32_PERF_STATUS](https://github.com/openhardwaremonitor/openhardwaremonitor/blob/e199e0ccc69b4da92495266ebc0faf7daad97608/Hardware/CPU/IntelCPU.cs#L408).  That's just not possible from within Java, even with JNA.

**Comment by dbwiddis** [comment]
OK, I trudged through the same links I visited about a year ago, but at least this time I'll take some notes on this issue to explain why it just can't be done.   

Windows offers an API that I linked above that I think is the best we can do.  It appears to report the same numbers as in the Registry at `HKLM\HARDWARE\DESCRIPTION\System\CentralProcessor\0` . WMI `Win32_Processor` also appears to provide this number, but only once.

Measuring time and getting clock cycles using `rdstc` works in normal situation but doesn't work with turbo-boost as the results are adjusted.  And It's unusable from Java, anyway.  Intel offers MSRs and we could read IA32_PERF_STATUS but it cannot be read from user mode, and requires a kernel driver (this is how OpenHardwareMonitor does it) and even then isn't doable in Java/JNA without custom JNI code.  The only option that I think *might* work is writing code guaranteed to execute exactly once every clock cycle and running it for some period of time. 

Then there's the philosophical answer: It's always changing, often in response to CPU load, so measuring by inducing load gets you in a situation like Heisenberg's uncertainty principle -- measuring the CPU frequency accuracy induces load on the CPU which changes the frequency which gives the result during your measurement but it's meaningless as soon as you're done.

In the end, OSHI reports what Windows has chosen to report, and [it's just not possible to do otherwise in a general, portable manner](https://stackoverflow.com/questions/8351944/finding-out-the-cpu-clock-frequency-per-core-per-processor).  

 I think I need to create a new label here for “I'm sorry Dave, I'm afraid I can't do that” ...

**Comment by dbwiddis** [comment]
OK, I wouldn't let this go away and after much research have found the solution, I think.  Windows 10 has a "Processer Performance Monitor" feature in Perfmon that includes actual frequency (or % of max).
![image](https://user-images.githubusercontent.com/9291703/64314571-e86de300-cf5b-11e9-9a22-b2ccdb3dad29.png)

It still jumps around so much one wonders how useful a snapshot is, but this is at least worth attempting to find the counter and use its results if it exists, falling back to the current method otherwise.

**Comment by dbwiddis** [comment]
I think I've cracked this nut!

```
Intel(R) Core(TM) i5-6300U CPU @ 2.40GHz
 1 physical CPU package(s)
 2 physical CPU core(s)
 4 logical CPU(s)
```
Reading from the above performance counters, there's one for each logical CPU:
```
2019-09-05T12:48:59.347 ProcessorFrequency: [2396, 2396, 1497, 1497] PercentofMaximumFrequency: [77, 77, 78, 78] 
2019-09-05T12:49:01.486 ProcessorFrequency: [2396, 2396, 1497, 2396] PercentofMaximumFrequency: [75, 74, 75, 75] 
2019-09-05T12:49:03.490 ProcessorFrequency: [2396, 1497, 1497, 2396] PercentofMaximumFrequency: [65, 65, 66, 66] 
2019-09-05T12:49:05.543 ProcessorFrequency: [2396, 2396, 2396, 1497] PercentofMaximumFrequency: [81, 81, 81, 82] 
2019-09-05T12:49:07.548 ProcessorFrequency: [2396, 2396, 2396, 1497] PercentofMaximumFrequency: [70, 70, 70, 70] 
2019-09-05T12:49:09.552 ProcessorFrequency: [1497, 2396, 1497, 1497] PercentofMaximumFrequency: [60, 60, 60, 60] 
2019-09-05T12:49:11.555 ProcessorFrequency: [2396, 1497, 2396, 2396] PercentofMaximumFrequency: [63, 64, 66, 67] 
2019-09-05T12:49:13.558 ProcessorFrequency: [2396, 2396, 2396, 2396] PercentofMaximumFrequency: [63, 68, 63, 64] 
2019-09-05T12:49:15.559 ProcessorFrequency: [1497, 2396, 1497, 1497] PercentofMaximumFrequency: [65, 64, 63, 64] 
2019-09-05T12:49:17.562 ProcessorFrequency: [1497, 2396, 2396, 1497] PercentofMaximumFrequency: [61, 65, 68, 63] 
2019-09-05T12:49:19.564 ProcessorFrequency: [1497, 2396, 2396, 2396] PercentofMaximumFrequency: [64, 60, 60, 60] 
2019-09-05T12:49:21.566 ProcessorFrequency: [1497, 2396, 1497, 1497] PercentofMaximumFrequency: [60, 60, 61, 60] 
2019-09-05T12:49:23.567 ProcessorFrequency: [1497, 1497, 1497, 2396] PercentofMaximumFrequency: [60, 59, 59, 59] 
2019-09-05T12:49:25.569 ProcessorFrequency: [2396, 2396, 2396, 2396] PercentofMaximumFrequency: [65, 63, 65, 62] 
2019-09-05T12:49:27.585 ProcessorFrequency: [2396, 2396, 1497, 2396] PercentofMaximumFrequency: [27, 27, 27, 27] 
2019-09-05T12:49:29.732 ProcessorFrequency: [2396, 2396, 2396, 2396] PercentofMaximumFrequency: [51, 51, 51, 51] 
2019-09-05T12:49:31.774 ProcessorFrequency: [2396, 2396, 2396, 2396] PercentofMaximumFrequency: [69, 69, 69, 69] 
2019-09-05T12:49:33.960 ProcessorFrequency: [2396, 2396, 2396, 2396] PercentofMaximumFrequency: [87, 87, 87, 87] 
2019-09-05T12:49:36.132 ProcessorFrequency: [2396, 2396, 2396, 2396] PercentofMaximumFrequency: [96, 96, 96, 96] 
2019-09-05T12:49:38.307 ProcessorFrequency: [2396, 2396, 2396, 2396] PercentofMaximumFrequency: [96, 96, 96, 96] 
```
So I'll update OSHI to use these counters if available (Should be on Win10+) or use WMI version of same data (Win7+ but doesn't look like it changes).

**Comment by dbwiddis** [comment]
And the plot thickens.  I think OSHI is already reporting the same value for Processor Frequency.  Reviewing the list in the previous comment, I noted there was not a direct correlation between (per-)processor frequency and percent of max.  The (logical processor) frequencies alternate between 2.4 GHz and 1.5 GHz.... this exactly matches what OSHI currently reports with the native `CallNtPowerInformation` call.  Here's a 1-second polling (I had to unplug to generate the low power mode that prompted this):
```
Current Frequencies: 2.4 GHz, 2.4 GHz, 2.4 GHz, 2.4 GHz
Current Frequencies: 1.5 GHz, 1.5 GHz, 2.4 GHz, 2.4 GHz
Current Frequencies: 2.4 GHz, 2.4 GHz, 2.4 GHz, 2.4 GHz
Current Frequencies: 2.4 GHz, 2.4 GHz, 2.4 GHz, 1.5 GHz
Current Frequencies: 2.4 GHz, 2.4 GHz, 1.5 GHz, 2.4 GHz
Current Frequencies: 1.5 GHz, 2.4 GHz, 2.4 GHz, 2.4 GHz
Current Frequencies: 2.4 GHz, 2.4 GHz, 2.4 GHz, 1.5 GHz
Current Frequencies: 2.4 GHz, 2.4 GHz, 1.5 GHz, 1.5 GHz
Current Frequencies: 1.5 GHz, 2.4 GHz, 2.4 GHz, 2.4 GHz
Current Frequencies: 1.5 GHz, 2.4 GHz, 1.5 GHz, 1.5 GHz
Current Frequencies: 2.4 GHz, 2.4 GHz, 1.5 GHz, 1.5 GHz
Current Frequencies: 1.5 GHz, 2.4 GHz, 1.5 GHz, 1.5 GHz
```
Which goes back to one of the "logical vs. physical" conversation.  I have one physical package and it doesn't really change frequency; what happens is the logical processors are sometimes throttled to 62.5% of their throughput, but not all at once.  If I only do 1 of 4 processors the overall average is 90%.  If I throttle 2 of 4 it's 81%, etc.

Which is what the "percent of maximum" appears to try to measure (but it does so across all the processors).  I suspect internally it says (hey, I need 80%  of the frequency) so the CPU decides to throttle 2 logical processors.... which two are likely dependent on currently executing threads.

Now, the numbers don't line up exactly, probably because the "on/off" nature of the logical processor throttle turns it into a game of probability whether you happen to catch it on it's "off" cycle.

So, in summary, I'm not sure what to do.  I think there's room to add in that "percent" value because it seems a slower-moving target and additional information that we don't have, and could probably be returned to the user if they wanted to manually calculate vs. the max.  Or maybe display it just as it is, a percentage, like on the task manager.
![image](https://user-images.githubusercontent.com/9291703/64388755-88734d00-cfec-11e9-84d7-a6835991937c.png)

**Comment by dtaskoff** [comment]
I thought I might share some experience from doing the same thing on another project:
on some machines, `PercentofMaximumFrequency` might always return the same value, however `PercentProcessorPerformance` seems to always give the accurate percentage, or at least
```
Win32_Processor.MaxClockSpeed
  * (Win32_PerfFormattedDataCounters_ProcessorInformation.PercentProcessorPerformance / 100)
```
matches Task Manager's output in the Performance tab.
